docs/website/docs/v1.x/dsl/middleware.md

```diff
@@ -3,4 +3,422 @@ sidebar_position: "8"
+                _    <- MyAuthService.doAuth(request)
+                // log request
+                _    <- logRequest(request)
+                // core business logic
+                user <- dbService.lookupUsersById(id).map(Response.json(_.json))
+                resp <- Response.json(user.toJson)
+                // log response
+                _    <- logResponse(resp)                
+            } yield resp)
+                    .timeout(2.seconds)
+                    .retryN(5)
+```
+Imagine repeating this for all our end points!!!
+
+So there are two problems with this approach
+* We are dangerously coupling our business logic with a lower level concern (like applying timeouts)
+* Also, we will have to do it for every single route in the system. For 100 routes we will need to repeat 100 timeouts!!!
+
+This can lead to a lot of boilerplate clogging our neatly written endpoints affecting readability, thereby leading to maintenance costs.
+
+## Need for middlewares and handling "aspects"
+
+If we refer to Wikipedia for the definition of an "[Aspect](https://en.wikipedia.org/wiki/Aspect_(computer_programming))" we can glean the following points.
+
+* An aspect of a program is a feature linked to many other parts of the program (**_most common example logging_**)., 
+* But it is not related to the program's primary function (**_core business logic_**) 
+* An aspect crosscuts the program's core concerns (**_for example logging code intertwined with core business logic_**),  
+* Therefore, it can violate the principle of "separation of concerns" which tries to encapsulate unrelated functions. (**_Code duplication and maintenance nightmare_**)
+
+Or in short, aspect is a common concern required throughout the application, and its implementation could lead to repeated boilerplate code and in violation of the principle of separation of concerns.
+There is a paradigm in the programming world called [aspect-oriented programming](https://en.wikipedia.org/wiki/Aspect-oriented_programming) that aims for modular handling of these common concerns in an application. 
+
+Some examples of common "aspects" required throughout the application
+- logging,
+- timeouts (preventing long-running code)
+- retries (or handling flakiness for example while accessing third party APIs)
+- authenticating a user before using the REST resource (basic, or custom ones like OAuth / single sign-on, etc).
+
+This is where middleware comes to the rescue. 
+Using middlewares we can compose out-of-the-box middlewares (or our custom middlewares) to address the above-mentioned concerns using ++ and @@ operators as shown below.
+
+#### Cleaned up code using middleware to address cross-cutting concerns like auth, request/response logging, etc.
+```scala
+// compose basic auth, request/response logging, timeouts middlewares
+val composedMiddlewares = Middleware.basicAuth("user","pw") ++ 
+        Middleware.debug ++ 
+        Middleware.timeout(5 seconds) 
```

Comment:
can describe what are we doing here. And how each piece relates to the work we were doing before.